### PR TITLE
[python] bring pandas support to the sklearn wrapper back

### LINKED ISSUE #901: sklearn categorical handling
I try to use the scikit-learn api for categorical handling in lightgbm with the latest master branch:

```
import pandas as pd
import lightgbm as lgb

df_train = pd.DataFrame({'target':[0,1,0,0],
                         'A':['a','b','c','a'],
                        'B':[1,2,2,3]})
df_train.A = df_train.A.astype('category')
df_test = pd.DataFrame({'target':[0,1,1,0],
                         'A':['a','b','c','unseed'],
                       'B':[1,2,2,3]})
df_test.A = df_test.A.astype('category')

classifier = lgb.LGBMClassifier()

X_train = df_train.drop(['target'], axis=1)
X_test = df_test.drop(['target'], axis=1)

y_train = df_train.target
y_est = df_test.target

classifier.fit(X_train, y_train, categorical_feature='auto')
```
The column A which is a pandas.Categorical should be treated automatically according to API documentation as categorical_feature='auto' is set.

However, I only get an error of:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-17-6e7c162874d1> in <module>()
     19 y_est = df_test.target
     20 
---> 21 classifier.fit(X_train, y_train, categorical_feature='auto') # should read pandas categories automatically ??

/usr/local/lib/python3.6/site-packages/lightgbm/sklearn.py in fit(self, X, y, sample_weight, init_score, eval_set, eval_names, eval_sample_weight, eval_init_score, eval_metric, early_stopping_rounds, verbose, feature_name, categorical_feature, callbacks)
    655                                         verbose=verbose, feature_name=feature_name,
    656                                         categorical_feature=categorical_feature,
--> 657                                         callbacks=callbacks)
    658         return self
    659 

/usr/local/lib/python3.6/site-packages/lightgbm/sklearn.py in fit(self, X, y, sample_weight, init_score, group, eval_set, eval_names, eval_sample_weight, eval_init_score, eval_group, eval_metric, early_stopping_rounds, verbose, feature_name, categorical_feature, callbacks)
    408             params['metric'] = eval_metric
    409 
--> 410         X, y = _LGBMCheckXY(X, y, accept_sparse=True, force_all_finite=False, ensure_min_samples=2)
    411         _LGBMCheckConsistentLength(X, y, sample_weight)
    412         self._n_features = X.shape[1]

/usr/local/lib/python3.6/site-packages/sklearn/utils/validation.py in check_X_y(X, y, accept_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, multi_output, ensure_min_samples, ensure_min_features, y_numeric, warn_on_dtype, estimator)
    540     X = check_array(X, accept_sparse, dtype, order, copy, force_all_finite,
    541                     ensure_2d, allow_nd, ensure_min_samples,
--> 542                     ensure_min_features, warn_on_dtype, estimator)
    543     if multi_output:
    544         y = check_array(y, 'csr', force_all_finite=True, ensure_2d=False,

/usr/local/lib/python3.6/site-packages/sklearn/utils/validation.py in check_array(array, accept_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, warn_on_dtype, estimator)
    415         # make sure we actually converted to numeric:
    416         if dtype_numeric and array.dtype.kind == "O":
--> 417             array = array.astype(np.float64)
    418         if not allow_nd and array.ndim >= 3:
    419             raise ValueError("Found array with dim %d. %s expected <= 2."

ValueError: could not convert string to float: 'a'

```
which indicates that the categories are not converted automatically.

### PR DESCRIPTION
**Pull request by StrikerRUS**
My fault is that in recent [PR](https://github.com/Microsoft/LightGBM/pull/870 ) with sklearn compatibility I forgot about pandas support. This PR is expected to bring pandas support back to the sklearn wrapper.
Fix #901 .

Since scikit-learn has poor pandas support, I think that the best thing to do is just pass `pandas.DataFrame` to Booster as is.

### CODE REVIEW

**Comment by wxchan** [comment]
aren't there too many tests now?

**Comment by guolinke** [comment]
@wxchan @StrikerRUS 
I think the python version 2.7 3.4 3.5 of OXS test can be removed.
For the appveyor, can we only keep one vs and one mingw ?

**Comment by wxchan** [comment]
@guolinke can we disable if-else test on osx?

**Comment by guolinke** [comment]
@wxchan yeah

**Comment by StrikerRUS** [comment]
@guolinke @wxchan The current total runtime of Appveyor and Travis is approximately equal (if there are no issues on Travis side :) ) that I think is reasonable. In addition, Travis jobs run in parallel mode (practically all osx-jobs starts at the same time), so it will be no significant reduction of the runtime.
Also I want to remind that the bug of overriding parameters was found with the help of random test (OSX Python 3.5 which you want to remove).

In my opinion, if-else case on OSX could be disabled but all other tests are important.
@guolinke Why you want to remove them? It the runtime inadmissible?

**Comment by guolinke** [comment]
@StrikerRUS 
The reason is the osx test are slow and easily time-out.

**Comment by StrikerRUS** [comment]
@guolinke I believe that it was a sporadic case of Travis OSX freezes.

File: tests/python_package_test/test_sklearn.py [hunk 1]
```diff
@@ -194,3 +199,33 @@ def test_sklearn_integration(self):
...
+        pred2 = list(gbm2.predict(X_test))
+        gbm3 = lgb.sklearn.LGBMClassifier().fit(X, y, categorical_feature=['A', 'B', 'C', 'D'])
+        pred3 = list(gbm3.predict(X_test))
+        gbm3.booster_.save_model('categorical.model')
+        gbm4 = lgb.Booster(model_file='categorical.model')
+        pred4 = list(gbm4.predict(X_test))
+        np.testing.assert_almost_equal(pred0, pred1)
+        np.testing.assert_almost_equal(pred0, pred2)
+        np.testing.assert_almost_equal(pred0, pred3)
+        np.testing.assert_almost_equal(pred0, pred4)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py]: @wxchan I can't find out what's wrong with the last test. Can you please help me?

Comment (henry0312) [reply tests/python_package_test/test_sklearn.py]: I guess this is because `gbm4` isn't regarded as Classifier.

**Comment by henry0312** [comment]
I think it's better to run tests with Python 2.7, 3.4, 3.5 and 3.6 and we should run the regular test with Python 2.7.x and 3.6.x at least. > https://github.com/Microsoft/LightGBM/pull/904#issuecomment-328403749, https://github.com/Microsoft/LightGBM/pull/904#issuecomment-328468317

Additionaly, as @StrikerRUS pointed out, if we remove some tests of OSX, the total time of tests won't change.

File: tests/python_package_test/test_sklearn.py (hunk 1 above)
Comment (StrikerRUS) [reply tests/python_package_test/test_sklearn.py]: @henry0312 Of course, it's probabilities! How could I be so inattentive?..
Thank you very much!

**Comment by henry0312** [comment]
@guolinke please run travis ci manually. there was something wrong.

**Comment by guolinke** [comment]
any updates of this ? ready to merge ?

**Pull request merged**